File: Sources/BloomCore/Remote/Bloomd.swift
import Foundation

/// The Python file Bloom puts on a server, and everything about keeping it current.
///
/// **How it is versioned: the file says so, in one literal, and that literal is the only thing
/// compared.** `bloomd version` prints it, `Bloomd.version(of:)` reads it out of the copy inside
/// the app bundle, and an install happens whenever the two strings differ. Not a hash of the file,
/// because reformatting a comment would then reinstall on every server the user has; not a
/// modification date, because a file copied by `rsync` or restored from a backup carries somebody
/// else's; not "newer than", because a version is a string and Bloom only ever ships forward, so
/// "different" is the whole of the question. Bumping the literal in `Resources/bloomd.py` is what
/// deploys it, and there is nothing else to remember.
public enum Bloomd {
    /// Where it goes on the server, relative to that user's home.
    ///
    /// Under the home directory rather than `/usr/local/bin`, because installing there needs
    /// `sudo` and Bloom has no business asking for it: everything here runs as the user whose SSH
    /// key opened the connection, and a per-user install is the honest scope for a per-user tool.
    /// The directory does not exist on a fresh server, which is why `install` creates it.
    public static let relativePath = ".bloom/bin/bloomd"

    /// The absolute path on a server whose home directory is `home`.
    ///
    /// An absolute path rather than `~/...` because the path is quoted for the remote shell, and
    /// quoting is exactly what stops `~` expanding. The home directory is a fact the probe already
    /// brought back, so there is nothing to guess.
    public static func path(inHome home: String) -> String {
        (home as NSString).appendingPathComponent(relativePath)
    }

    /// The version literal at the top of a `bloomd.py`.
    ///
    /// Deliberately a small scan rather than a regular expression over the whole file: the literal
    /// is on one line, in one form, and a scanner that reads the first assignment and stops cannot
    /// be fooled by the same words appearing in the docstring below it.
    public static func version(of source: String) -> String? {
        for line in source.components(separatedBy: .newlines) {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            guard trimmed.hasPrefix("BLOOMD_VERSION") else { continue }
            guard let open = trimmed.firstIndex(of: "\""),
                  let close = trimmed.lastIndex(of: "\""),
                  open < close else { return nil }
            let value = String(trimmed[trimmed.index(after: open)..<close])
            return value.isEmpty ? nil : value
        }
        return nil
    }

    /// The copy inside the running app bundle.
    ///
    /// The same shape as `BridgeRegistration.shimPath`, and for the same two reasons: it is
    /// resolved from the executable rather than through `Bundle.main.url(forResource:)` so that a
    /// pure function can be tested against a made-up layout, and there is an environment override
    /// because `Tools/test-core.sh` mirrors only BloomCore into a package with no app target, so
    /// nothing the suite builds has a bundle to look in.
    public static func sourcePath(
        beside executable: String? = Bundle.main.executableURL?.path,
        environment: [String: String] = ProcessInfo.processInfo.environment
    ) -> String? {
        if let override = environment["BLOOM_BLOOMD_PATH"], !override.isEmpty {
            return FileManager.default.isReadableFile(atPath: override) ? override : nil
        }
        guard let executable else { return nil }
        // Contents/MacOS/Bloom, so the resource is one directory up and along.
        let macos = (executable as NSString).deletingLastPathComponent
        let contents = (macos as NSString).deletingLastPathComponent
        let candidate = (contents as NSString).appendingPathComponent("Resources/bloomd.py")
        return FileManager.default.isReadableFile(atPath: candidate) ? candidate : nil
    }
}

/// What the daemon on a server needs doing to it, worked out before anything is copied.
///
/// A value rather than a branch inside the install, so the pane can say what it is about to do and
/// so the decision is testable without a server. "Copy that file, so it works seamless" is this
/// enum: nobody is ever asked whether to install, and the only case that is not automatic is the
/// one where a copy would not help.
public enum BloomdAction: Sendable, Hashable {
    case upToDate(version: String)
    case install(reason: BloomdReason, version: String)
    /// The server has no Python 3, so there is nothing that could run the file once it arrived.
    case impossible

    public var needsCopying: Bool {
        if case .install = self { return true }
        return false
    }
}

public enum BloomdReason: Sendable, Hashable {
    case notThere
    case differentVersion(String)
    /// The file is there and did not answer with a version, which is a truncated copy or a Python
    /// that could not parse it. Either way the answer is the same file again.
    case unreadable

    public var sentence: String {
        switch self {
        case .notThere: "not installed yet"
        case .differentVersion(let found): "version \(found) is installed"
        case .unreadable: "the installed copy did not answer"
        }
    }
}

public extension Bloomd {
    /// Compares what the bundle ships with what the server reported.
    ///
    /// `hasPython` is a parameter rather than something looked up here, because the probe already
    /// asked and this has to stay a pure function: the whole decision is four lines and the value
    /// of testing it is that nobody has to run a server to find out what it does.
    static func decide(
        shipping local: String,
        installed remote: String?,
        hasPython: Bool
    ) -> BloomdAction {
        guard hasPython else { return .impossible }
        guard let remote, !remote.isEmpty else {
            return .install(reason: .notThere, version: local)
        }
        guard remote == local else {
            return .install(reason: .differentVersion(remote), version: local)
        }
        return .upToDate(version: local)
    }
}

// MARK: - Talking to it

/// Runs `bloomd` on a server, whatever "on a server" turns out to mean.
///
/// It holds a `CommandRunning` rather than an SSH connection, so the same type drives the copy on
/// this Mac. That is not a test convenience: it is how the local and the remote diff pass can
/// eventually be the same code with a different runner behind it.
public struct BloomdClient: Sendable {
    private let runner: any CommandRunning
    /// The remote home directory, from the probe. Everything is addressed from it.
    private let home: String

    public init(runner: any CommandRunning, home: String) {
        self.runner = runner
        self.home = home
    }

    public var remotePath: String { Bloomd.path(inHome: home) }

    /// Copies the file over and checks that the copy answers.
    ///
    /// **The check is the point of the second half.** A `cat` that wrote zero bytes exits zero,
    /// and a server that has just been handed an empty file would then be recorded as running the
    /// version Bloom meant to install. So the version is asked for afterwards, from the file that
    /// is actually there, and a mismatch is a failure rather than a hopeful log line.
    @discardableResult
    public func install(source: String, timeout: Duration = .seconds(30)) async throws -> String {
        try await runner.put(source, at: remotePath, executable: true, timeout: timeout)
        guard let installed = try await version(timeout: timeout) else {
            throw BloomdTrouble.silentAfterInstall
        }
        guard installed == Bloomd.version(of: source) else {
            throw BloomdTrouble.wrongVersionAfterInstall(installed)
        }
        return installed
    }

    /// What the copy on the server says it is, or nil when there is nothing there.
    public func version(timeout: Duration = .seconds(15)) async throws -> String? {
        let result = try await runner.run(command(["version"], timeout: timeout))
        guard result.ok else { return nil }
        let text = result.trimmed
        return text.isEmpty ? nil : text
    }

    /// The several git facts one diff pass needs, in one round trip.
    public func status(
        worktree: String,
        base: String,
        timeout: Duration = .seconds(60)
    ) async throws -> BloomdStatus {
        let result = try await runner.run(
            command(["status", worktree, "--base", base], timeout: timeout)
        )
        return try BloomdStatus.decode(result.stdout, status: result.status)
    }

    /// `python3 <path> <arguments>`, as argv.
    ///
    /// `python3` explicitly rather than relying on the shebang, because the file is installed
    /// under a home directory and a home directory is exactly the kind of mount that gets `noexec`
    /// on a hardened server. Running the interpreter on the file needs no execute bit at all.
    private func command(_ arguments: [String], timeout: Duration) -> CommandLaunch {
        CommandLaunch(
            executable: "python3",
            arguments: [remotePath] + arguments,
            timeout: timeout
        )
    }
}

public enum BloomdTrouble: Error, Sendable, Hashable, CustomStringConvertible {
    case silentAfterInstall
    case wrongVersionAfterInstall(String)
    case refused(String)
    case unreadable(String)

    public var description: String {
        switch self {
        case .silentAfterInstall:
            "The file was copied and then said nothing when it was run. Check that python3 works on that server."
        case .wrongVersionAfterInstall(let found):
            "The file was copied and the server is still running version \(found)."
        case .refused(let message):
            message
        case .unreadable(let head):
            "bloomd answered with something that is not JSON: \(head)"
        }
    }
}

/// One worktree, as the server sees it.
public struct BloomdStatus: Sendable, Hashable, Codable {
    public var version: String
    public var path: String
    public var base: String
    /// Nil when the base branch is not on that server, in which case the counts below are
    /// uncommitted work only. A caller that draws "12 files changed" wants to know which of the
    /// two questions it answered.
    public var mergeBase: String?
    public var branch: String?
    public var head: String?
    public var upstream: String?
    public var ahead: Int
    public var behind: Int
    public var changedFiles: Int
    public var additions: Int
    public var deletions: Int
    public var dirty: Bool
    public var files: [BloomdFile]
}

public struct BloomdFile: Sendable, Hashable, Codable {
    public var path: String
    public var oldPath: String?
    public var change: String
    public var additions: Int
    public var deletions: Int
    public var binary: Bool
}

extension BloomdStatus {
    /// Reads one answer.
    ///
    /// The failure blob is decoded before the success one, because `bloomd` reports a bad worktree
    /// as `{"ok": false, "error": ...}` with exit status 1, and that message is the useful thing
    /// on the screen. Anything that is not either shape is quoted back verbatim rather than
    /// summarised, since output nobody anticipated is exactly what somebody debugging needs to see.
    static func decode(_ output: String, status: Int32) throws -> BloomdStatus {
        let data = Data(output.utf8)
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase

        if let refusal = try? decoder.decode(BloomdRefusal.self, from: data), !refusal.ok {
            throw BloomdTrouble.refused(refusal.error)
        }
        do {
            return try decoder.decode(BloomdStatus.self, from: data)
        } catch {
            let head = output.trimmingCharacters(in: .whitespacesAndNewlines).prefix(200)
            throw BloomdTrouble.unreadable(
                head.isEmpty ? "nothing, and it exited \(status)" : String(head)
            )
        }
    }
}

private struct BloomdRefusal: Decodable {
    let ok: Bool
    let error: String
}

File: Resources/bloomd.py
#!/usr/bin/env python3
"""bloomd: the half of Bloom that lives on your server.

Bloom is a Mac app that runs coding agents in git worktrees. When a workspace lives on a server
rather than on the laptop, Bloom needs answers about that worktree, and asking for them one `git`
at a time over SSH is a network round trip per question: the diff pass alone is over a hundred
processes, and six sequential questions at 20ms each is longer than the work itself. This file is
what makes it one question instead.

Three things it is, and each of them on purpose:

  * **One file, Python 3, no third-party imports.** Python 3 is on every mainstream server
    distribution, so installing this is copying a file. Nothing to build, nothing to `pip install`,
    nothing to keep up to date but the file itself.
  * **No network listener.** It has no port, no socket and no daemon mode, despite the name. Bloom
    speaks to it over the SSH connection it already has: stdin in, stdout out, one process per
    call. Handing an application SSH access to your server is a decision you already took; handing
    it permission to open a port is a different and larger one, and this asks for nothing extra.
  * **Readable.** Somebody who lets an app copy a file onto their server should be able to `cat`
    it and see the whole of what it does. That is why it is Python rather than a compiled binary,
    and it is why this docstring is here.

Usage:

    python3 bloomd version
    python3 bloomd status <worktree> [--base <branch>]

Every command writes one JSON object to stdout and exits 0, or writes one JSON object with
"ok": false and exits 1. `version` is the exception and prints a bare version string, because it
is what an installer compares and a bare string needs no parser.

**What is deliberately not here yet: `run` and `follow`.** Starting an agent on the server and
streaming its output back is the next thing this file grows, and the dispatch table at the foot is
shaped for it: a command is a function taking parsed arguments, and a streaming one will write
newline delimited JSON to stdout and flush after every line rather than returning a blob. Nothing
above needs to change for that to arrive.

This file writes nothing outside the worktree it is pointed at, and reads no credential of any
kind.
"""

import json
import os
import subprocess
import sys

# The one fact an installer compares. Bloom reads this literal out of its own copy of this file,
# runs `bloomd version` on the server, and copies when the two differ. It is a bare string rather
# than a tuple so that both sides compare exactly what is printed.
BLOOMD_VERSION = "1"

# Long enough for a cold index on a large repository, short enough that a wedged git does not hold
# an SSH channel open until Bloom's own timeout kills the connection underneath it.
GIT_TIMEOUT_SECONDS = 60


class Failure(Exception):
    """Something went wrong that the caller should be told about in words."""


def git(worktree, args):
    """Run one git command in a worktree and hand back its stdout as bytes.

    Bytes, not text, and this is not fastidiousness. Git's default output C-quotes any path that
    is not plain ASCII, so `cafe.txt` with an accent comes back as an escaped ASCII string that no
    longer names the file. Every caller below passes `-z`, which turns the quoting off and makes
    NUL the separator, and NUL is the one byte a path cannot contain. Decoding happens once, at
    the end, with surrogateescape, so a path that is not valid UTF-8 survives the trip instead of
    raising.
    """
    command = ["git", "-C", worktree] + list(args)
    environment = dict(os.environ)
    # The diff pass is a reader and must never wait on a writer. Without this, a `git status` run
    # while the user has an editor open takes the index lock and blocks.
    environment["GIT_OPTIONAL_LOCKS"] = "0"
    # No pager, ever: a pager attached to a pipe is a hang.
    environment["GIT_PAGER"] = "cat"
    environment["LC_ALL"] = "C"

    try:
        finished = subprocess.run(
            command,
            env=environment,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            timeout=GIT_TIMEOUT_SECONDS,
        )
    except FileNotFoundError:
        raise Failure("git is not installed on this server")
    except subprocess.TimeoutExpired:
        raise Failure("git took longer than %d seconds" % GIT_TIMEOUT_SECONDS)

    if finished.returncode != 0:
        message = decode(finished.stderr).strip() or "git exited %d" % finished.returncode
        raise Failure(message)
    return finished.stdout


def decode(raw):
    """Bytes to text without ever raising. See `git` for why the bytes were kept that long."""
    return raw.decode("utf-8", "surrogateescape")


def records(raw):
    """Split `-z` output into its NUL separated records, dropping the empty tail."""
    parts = raw.split(b"\0")
    if parts and parts[-1] == b"":
        parts.pop()
    return parts


# ---------------------------------------------------------------------------
# status
# ---------------------------------------------------------------------------

def branch_facts(worktree):
    """Head, upstream and how far apart they are, from one `git status`.

    `--porcelain=v2 --branch -z` is asked for rather than three `rev-parse` calls because it is
    one process for facts that must agree with each other: a head read separately from an upstream
    can describe two different moments if a fetch lands between them.
    """
    raw = git(worktree, ["status", "--porcelain=v2", "--branch", "-z"])
    facts = {"head": None, "branch": None, "upstream": None, "ahead": 0, "behind": 0}
    for record in records(raw):
        line = decode(record)
        if not line.startswith("# branch."):
            continue
        key, _, value = line[len("# branch."):].partition(" ")
        if key == "oid" and value != "(initial)":
            facts["head"] = value
        elif key == "head" and value != "(detached)":
            facts["branch"] = value
        elif key == "upstream":
            facts["upstream"] = value
        elif key == "ab":
            # "+3 -1", and either half can be absent on a malformed line.
            for token in value.split():
                if token.startswith("+"):
                    facts["ahead"] = int(token[1:] or 0)
                elif token.startswith("-"):
                    facts["behind"] = int(token[1:] or 0)
    return facts


def merge_base(worktree, base):
    """Where this branch left `base`, or None when there is no such branch here.

    None rather than an exception, because a workspace whose base branch has been deleted is a
    thing that happens and is not a reason for the whole status call to fail. The caller falls
    back to HEAD, which measures uncommitted work only, and says so in the answer.
    """
    for candidate in (base, "origin/" + base):
        try:
            return decode(git(worktree, ["merge-base", "HEAD", candidate])).strip()
        except Failure:
            continue
    return None


def name_status(worktree, since):
    """What happened to each path: added, modified, deleted, renamed, copied.

    `R` and `C` glue a similarity score to the letter and are followed by TWO paths, old then new,
    which is the whole reason this is parsed record by record rather than split on a separator.
    """
    raw = git(worktree, ["diff", "--name-status", "-M", "-z", since, "--"])
    changes = {}
    queue = records(raw)
    index = 0
    while index < len(queue):
        code = decode(queue[index])[:1]
        index += 1
        if code in ("R", "C") and index + 1 < len(queue):
            old = decode(queue[index])
            new = decode(queue[index + 1])
            index += 2
            changes[new] = {"change": "renamed" if code == "R" else "copied", "old_path": old}
        elif index < len(queue):
            path = decode(queue[index])
            index += 1
            changes[path] = {"change": CHANGE_LETTERS.get(code, "modified"), "old_path": None}
        else:
            break
    return changes


CHANGE_LETTERS = {
    "A": "added",
    "M": "modified",
    "D": "deleted",
    "T": "modified",
}


def numstat(worktree, since, changes):
    """Additions and deletions per path.

    A record is `additions TAB deletions TAB path`, except for a rename or a copy, where the path
    field is empty and the old and new paths follow as two records of their own. A binary file
    reports "-" for both counts, which is a fact worth keeping rather than a zero to be silently
    counted.

    Only the first two tabs are separators. A path is allowed to contain one, so the split is
    bounded.
    """
    raw = git(worktree, ["diff", "--numstat", "-M", "-z", since, "--"])
    files = {}
    queue = records(raw)
    index = 0
    while index < len(queue):
        fields = queue[index].split(b"\t", 2)
        index += 1
        if len(fields) != 3:
            continue
        additions = decode(fields[0])
        deletions = decode(fields[1])
        path = decode(fields[2])
        old_path = None
        if fields[2] == b"":
            if index + 1 >= len(queue):
                break
            old_path = decode(queue[index])
            path = decode(queue[index + 1])
            index += 2

        recorded = changes.get(path, {})
        files[path] = {
            "path": path,
            "old_path": recorded.get("old_path") or old_path,
            "change": recorded.get("change") or ("modified" if old_path is None else "renamed"),
            "additions": 0 if additions == "-" else int(additions or 0),
            "deletions": 0 if deletions == "-" else int(deletions or 0),
            "binary": additions == "-",
        }
    return files


def untracked(worktree):
    """Files git has never seen, counted the way git counts a line.

    They belong in the answer whatever it is measured from: a file git does not know about is
    uncommitted against every commit. Every line is a line, and a trailing newline terminates the
    last one rather than starting an empty one, which is the off-by-one that made every untracked
    file read one addition too many the first time Bloom counted them.
    """
    raw = git(worktree, ["ls-files", "--others", "--exclude-standard", "-z"])
    files = {}
    for record in records(raw):
        path = decode(record)
        if not path:
            continue
        full = os.path.join(worktree, path)
        lines = 0
        binary = False
        try:
            with open(full, "rb") as handle:
                content = handle.read()
            if b"\0" in content:
                binary = True
            else:
                lines = content.count(b"\n")
                if content and not content.endswith(b"\n"):
                    lines += 1
        except OSError:
            # A file that vanished between the listing and the read, or one this user cannot open.
            # It is still a changed file, and its line count is the only thing not known.
            binary = True
        files[path] = {
            "path": path,
            "old_path": None,
            "change": "untracked",
            "additions": lines,
            "deletions": 0,
            "binary": binary,
        }
    return files


def command_status(arguments):
    """Everything one diff pass needs about one worktree, in one JSON object.

    This is the whole reason bloomd exists. Bloom's local pass is a merge base, three diffs and a
    line count per untracked file, which is over a hundred processes across a window full of
    workspaces. Over SSH that is over a hundred round trips; here it is one, and the processes are
    started on the machine the repository is already on.
    """
    if not arguments:
        raise Failure("status needs a worktree path")

    worktree = arguments[0]
    base = "main"
    rest = arguments[1:]
    while rest:
        if rest[0] == "--base" and len(rest) > 1:
            base = rest[1]
            rest = rest[2:]
        else:
            raise Failure("unknown argument: %s" % rest[0])

    if not os.path.isdir(worktree):
        raise Failure("no such directory: %s" % worktree)

    facts = branch_facts(worktree)
    since = merge_base(worktree, base)
    changes = name_status(worktree, since or "HEAD")
    files = numstat(worktree, since or "HEAD", changes)
    files.update(untracked(worktree))

    listing = sorted(files.values(), key=lambda entry: entry["path"])
    return {
        "ok": True,
        "version": BLOOMD_VERSION,
        "path": worktree,
        "base": base,
        # Nil when the base branch is not on this server, which is the case the caller has to be
        # able to tell apart: the numbers below are then uncommitted work only.
        "merge_base": since,
        "branch": facts["branch"],
        "head": facts["head"],
        "upstream": facts["upstream"],
        "ahead": facts["ahead"],
        "behind": facts["behind"],
        "changed_files": len(listing),
        "additions": sum(entry["additions"] for entry in listing),
        "deletions": sum(entry["deletions"] for entry in listing),
        "dirty": bool(listing),
        "files": listing,
    }


def command_version(arguments):
    """The version, bare, because an installer compares a string and not a document."""
    del arguments
    sys.stdout.write(BLOOMD_VERSION + "\n")
    return None


# The dispatch table, and the shape the next commands take. A command is a function of the
# arguments after its own name. Returning a dictionary means "print this as JSON and exit 0";
# returning None means the command has already written whatever it had to write, which is what
# `version` does and what a streaming `run` or `follow` will do, one JSON object per line, flushed
# as it goes.
COMMANDS = {
    "version": command_version,
    "status": command_status,
}


def main(argv):
    if not argv or argv[0] in ("-h", "--help", "help"):
        usage = __doc__.split("Usage:")[1].split("Every command")[0]
        for line in usage.strip().splitlines():
            sys.stdout.write(line.strip() + "\n")
        return 0 if argv else 2

    handler = COMMANDS.get(argv[0])
    if handler is None:
        json.dump({"ok": False, "error": "unknown command: %s" % argv[0]}, sys.stdout)
        sys.stdout.write("\n")
        return 1

    try:
        answer = handler(argv[1:])
    except Failure as failure:
        json.dump({"ok": False, "error": str(failure)}, sys.stdout)
        sys.stdout.write("\n")
        return 1

    if answer is not None:
        # Compact, because this crosses a network on every diff pass, and sorted so that two runs
        # over an unchanged worktree produce identical bytes.
        json.dump(answer, sys.stdout, separators=(",", ":"), sort_keys=True)
        sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))
